src/transform/src/predicate_pushdown.rs

```diff
@@ -459,95 +462,126 @@ impl PredicatePushdown {
+                                    ));
+                                }
+                            }
+                            None
+                        })
+                        .collect::<Vec<_>>();
+
+                    if let Some(literal_pos) = equivalence.iter().position(|expr| expr.is_literal())
+                    {
+                        // Case 2: There is one literal in the equivalence class
+                        let literal = equivalence[literal_pos].clone();
+                        let gen_literal_equality_pred = |expr| {
+                            if literal.is_literal_null() {
+                                MirScalarExpr::CallUnary {
+                                    func: expr::UnaryFunc::IsNull,
+                                    expr: Box::new(expr),
+                                }
+                            } else {
+                                MirScalarExpr::CallBinary {
+                                    func: expr::BinaryFunc::Eq,
+                                    expr1: Box::new(expr),
+                                    expr2: Box::new(literal.clone()),
+                                }
+                            }
+                        };
+                        // For every single-input expression `expr`, we can push
+                        // down `expr = literal` and remove `expr` from the
+                        // equivalence class.
+                        for (expr_pos, input, expr) in single_input_exprs.drain(..).rev() {
+                            push_downs[input].push(gen_literal_equality_pred(expr));
+                            equivalence.remove(expr_pos);
+                        }
+                        // If any multi-input expression `expr` exists in the
+                        // same equivalence class as the literal, the predicate
+                        // `expr = literal` should be
+                        // lifted above the join.
+                        for multi_input_expr in equivalence.drain(..) {
+                            lifts.push(gen_literal_equality_pred(multi_input_expr));
+                        }
+                    } else {
+                        // Case 3: There are no literals in the equivalence
+                        // class. For each single-input expression `expr1`,
+                        // scan the remaining single-input expressions
+                        // to see if there is another expression `expr2` from
+                        // the same input. If there is, push down
+                        // `(expr1 = expr2) || (isnull(expr1) &&
+                        // isnull(expr2))`. `expr1` can then
+                        // be removed from the equivalence class.
+                        while single_input_exprs.len() >= 2 {
+                            let (pos, input, expr1) = single_input_exprs.pop().unwrap();
+                            if let Some((_, _, expr2)) =
```

Comment:
Should this be a `while let`? It seems like you remove both `expr1` and `expr2`, but there could be an `expr3` as well.

Reply:
Changed to a `while let`.

Like I say the comment three lines above, I only remove `expr1`, not `expr2`. When I get around to popping `expr2` out of `single_input_exprs` is when I will check if `expr3` exists. If `expr3` exists, that's when I remove `expr2`.

The motivation is that I want to remove stuff from the equivalence class in right to left order so that each removal does not require me to recompute the positions of the things remaining to remove.

Reply:
Also, if I were to remove all exprs in an equivalence class belonging to an input, then I would be incorrectly removing the join variable constraints.

Reply:
Sorry, I misread `equivalence.remove(pos);` as removing `expr2` rather than `expr1`.